src/methods/potential_theory/methods.jl

```diff
@@ -0,0 +1,465 @@
+  5. Check convergence
+"""
+function solve_at_potential_fugacity(eos, P_init, z_init, T, z_bulk, f_bulk, P_bulk, ε_vec, alg; verbose=false)
+    
+    # Unpack
+    RT = Rg(eos) * T
+    n_comp = length(z_bulk)
+    
+    # Pre-compute constant term: C^i = z_g^i * f_g^i  * exp(ε^i/RT)
+    C = z_bulk .* f_bulk .* exp.(ε_vec ./ RT)
+    
+    # Tolerances
+    tol_z = alg.abstol
+    tol_P = alg.reltol
+    tol_stability = abs2(cbrt(tol_z))  # Stability threshold (Clapeyron pattern)
+    
+    # Initialize
+    P = P_init
+    z = copy(z_init)
+    
+    # OUTER LOOP: Newton iteration on pressure
+    for j in 1:alg.maxiter_outer
+        
+        # Checkpoint for stability restart
+        z_restart = copy(z)
+        P_restart = P
+        
+        valid_iter = true  # Flag for valid inner loop convergence
+        
+        # INNER LOOP: Successive substitution on composition
+        for i in 1:alg.maxiter_inner
+            
+            # Calculate fugacity coefficients at current (P, z)
+            # Using Clapeyron's lnϕ function
+            lnϕ_z = lnϕ(eos, P, T, z)[1]
+            ϕ_z = exp.(lnϕ_z)
+            f_z = ϕ_z .* P  # Fugacities
+            
+            # Update composition: z^i = C^i / f^i(P,z)
+            z_new = C ./ f_z
+            z_new = z_new / sum(z_new) # Normalize
+            
+            # Check convergence on composition
+            if norm(z_new - z, Inf) < tol_z
+                z = z_new
+                break
+            end
+            
+            #═════════════════════════════════════════════════════
+            # STABILITY CHECK (prevents oscillation/trivial solution)
+            #═════════════════════════════════════════════════════
+            stability = norm(z_new - z_bulk, Inf)  # ||z - z_bulk||∞
+            
+            if stability < tol_stability
+                # Segregated composition too similar to bulk
+                # → Iteration going wrong (weak field, wrong P, near critical)
+                
+                verbose && @warn "Stability check triggered at P=$P: ||z-z_bulk||=$stability < $tol_stability"
+                
+                # Restore to checkpoint
+                z = z_restart
+                P = P_restart
+                
+                # Mark iteration as invalid (skip derivative calc, force P update)
+                valid_iter = false
+                
+                # Exit inner loop
+                break
+            end
+            #═════════════════════════════════════════════════════
+            
+            # Update for next inner iteration
+            z = z_new
+        end
+        
+        # If stability check triggered, skip to pressure update
+        if !valid_iter
+            verbose && @info "Forcing pressure update due to stability check"
+        end
+        
+        # Pressure residual function for ForwardDiff
+        # F = Σ[C^i / f^i(P,z)] - 1 where f^i = ϕ^i * P
+        function F_pressure(P_trial)
```

Comment:
instead of doing this manually, there is `Clapeyron.∂lnϕ∂n∂P`. the magic of that function (and `Clapeyron.lnϕ` is that they accept an optional cache argument, generated by `Clapeyron.∂lnϕ_cache(model,p,T,z,needs_temperature_derivative::Val{true | false})`, that speeds up fugacity calculations a lot

Reply:
something like this?
```
f(p) = sum(C ./ (exp.(lnϕ) .* p )) - 1 = sum(  exp(-.lnϕ) .* 1/ p * C) - 1
dfdp = sum(Ci * dfidp)
dfidp = -exp(-.lnϕi) * dlnϕidp/p - exp(-lnϕi)/p^2
dfidp = (-exp(-.lnϕi)/p)*(dlnϕidp + 1/p)
```